### PR TITLE
FSDP2: Fix save/load, introduce CPU offload options, fix gradient checkpointing

### PR DESCRIPTION
**Pull request by bghira**
it wasn't properly selecting blocks to shard, and two FSDP options weren't exposed to the user, cpu offload and activation checkpointing.

we can't use torch's activation checkpointing with FSDP, we have to use FSDP-aware version. i think maybe this is FSDP2 limitation, perhaps - but it's faster to use the correct implementation anyway.

### CODE REVIEW

**Comment by bghira** [comment]
we can train Flux on A5000 with 24G, all modules and layers.

same for full-rank training of SD3.5 Large, and medium.

File: simpletuner/helpers/training/diffusers_overrides.py [hunk 1]
```diff
@@ -375,3 +380,65 @@ def enable_reversible_fusion():
...
+                    dist.broadcast(shard, src=0, group=dist.group.WORLD)
+                shards[name] = shard
+
+        model.load_state_dict(shards, assign=True)
+        return model
+
+    accelerate_fsdp_utils.fsdp2_load_full_state_dict = replacement
+
+
+print("Override FSDP2 state dict loader")
```
Comment (bghira) [inline simpletuner/helpers/training/diffusers_overrides.py]:
```suggestion
```
